music21/opensheetmusicdisplay/osmd.py

```diff
@@ -0,0 +1,189 @@
+                    console.log("loaded OSMD for the first time",opensheetmusicdisplay)
+                    resolve(opensheetmusicdisplay);
+                };
+                {{script_command}}
+                
+            }) 
+        }
+        loadOSMD().then((OSMD)=>{
+            console.log("loaded OSMD",OSMD)
+            var div_id = "{{DIV_ID}}";
+                console.log(div_id)
+            document.querySelector('#'+div_id).innerHTML = "";
+            window.openSheetMusicDisplay = new OSMD.OpenSheetMusicDisplay(div_id);
+            openSheetMusicDisplay
+                .load({{data}})
+                .then(
+                  function() {
+                    console.log("rendering data")
+                    openSheetMusicDisplay.render();
+                  }
+                );
+        })
+        """ \
+        .replace('{{DIV_ID}}',div_id) \
+        .replace('{{data}}',json.dumps(xml)) \
+        .replace('{{script_dir}}',pathname2url(script_dir))
+
+        if offline is True:
+            script_content = open(script_dir).read()
+            script = script.replace('{{script_command}}',"""
+                s.type = 'text/javascript';
+                s.text="""+json.dumps(script_content)+""";
+                document.body.appendChild( s ); // browser will try to load the new script tag
+                oncompleted();
+                """)
+        else:
+            script = script.replace('{{script_command}}',"""
+                s.setAttribute( 'src', 'https://cdn.jsdelivr.net/npm/opensheetmusicdisplay@0.3.1/build/opensheetmusicdisplay.min.js' );
+                s.onload=oncompleted;
+                document.body.appendChild( s ); // browser will try to load the new script tag
+                """)
+
+        return script
+
+    @staticmethod
+    def addDefaultPartName(score):
+        # If no partName is present in the first instrument, OSMD will display the ugly 'partId'
+        allInstruments = list(score.getInstruments(returnDefault=False, recurse=True))
+        
+        if len(allInstruments)==0:
+            # print("adding default inst")
+            defaultInstrument = Piano()
+            defaultInstrument.instrumentName = 'Default'
+            score.insert(None, defaultInstrument)
+            score.coreElementsChanged()
+        elif not allInstruments[0].instrumentName: 
+            # print("adding instrumentName")
+            # instrumentName must not be '' or None
+            allInstruments[0].instrumentName = 'Default'
+            score.coreElementsChanged()
+
+        return score
+
+class Test(unittest.TestCase):
+
+    @unittest.skipUnless(hasInstalledIPython, "skipping since IPython not installed")
+    def testOpenSheetMusicDisplayRuns(self):
+        from music21 import corpus, environment
+        environLocal = environment.Environment()
+
+        s = corpus.parse('luca/gloria').measures(1, 19)
+        # s.show('osmd')
+        ConverterOpenSheetMusicDisplay().show(s, None)
```

Comment:
this should be in TestExternal.  Test() only should be run if there are no side effects.  TestExternal is for anything that opens up a browser, plays music, etc.  Since the general test system runs all Test() functions, it isn't needed.

Reply:
Should be fixed now: 90779099fe7ffcebba6afe622e080a9ac8b2ebd5. Would be great to be able to check the output (e.g. a notebook that displays all the wigets and .js options?).